framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1210,71 +1217,137 @@ int pcm_drain(struct pcm *pcm)
 	}
 }
 
-static inline int pcm_mmap_avail(struct pcm *pcm)
-{
-	if (pcm == NULL) {
-		return 0;
-	}
-	return pcm_bytes_to_frames(pcm, pcm->next_size);
-}
-
+/** Application request to access a portion of direct (mmap) area
+ * @param[in] pcm A PCM handle
+ * @param[out] areas Returned mmap channel areas
+ * @param[out] offset Returned mmap area offset in frames
+ * @param[out] frames Mmap area portion size in frames
+ * @returns On success, zero is returned. On failure, a negative number returned
+ * @ingroup libtinyalsa-pcm
+ */
 int pcm_mmap_begin(struct pcm *pcm, void **areas, unsigned int *offset, unsigned int *frames)
 {
 	if (pcm == NULL || *areas == NULL || offset == NULL || frames == NULL) {
 		return -EINVAL;
 	}
-	int nframes = pcm_bytes_to_frames(pcm, pcm->next_size);
 
-	/* If data is not available, return -1 */
-	if (nframes == 0) {
+	int nframes = 0;
+	/* The next buffer to be used is pointed by mmap_idx */
+	struct ap_buffer_s *apb = pcm->pBuffers[pcm->mmap_idx];
+
+	/* If buffer is in enqueued state, return 0 */
+	if (apb->flags & AUDIO_APB_OUTPUT_ENQUEUED) {
 		*frames = 0;
-		return -1;
+		return 0;
 	}
 
-	*areas = pcm->next_buf->samp;
-	*offset = pcm->next_offset;
+	*areas = apb->samp;
 
-	if (*frames < nframes) {
-		pcm->next_size -= pcm_frames_to_bytes(pcm, *frames);
-		pcm->next_offset += pcm_frames_to_bytes(pcm, *frames);
+	if (pcm->flags & PCM_IN) {
+		*offset = pcm_bytes_to_frames(pcm, apb->curbyte);
+		nframes = pcm_bytes_to_frames(pcm, apb->nbytes - apb->curbyte);
 	} else {
-		*frames = pcm_bytes_to_frames(pcm, pcm->next_size);
-		pcm->next_size = 0;
-		pcm->next_offset = 0;
+		*offset = 0;
+		nframes = pcm_bytes_to_frames(pcm, apb->nmaxbytes);
+	}
+
+	if (*frames > nframes) {
+		*frames = nframes;
 	}
 
 	return 0;
 }
 
+/** Application request to access a portion of direct (mmap) area
```

Comment:
commit doing finish access & update position, right?

Reply:
True. 
But, I removed this description as we have defined it in header file. (as per your previous comment)